### PR TITLE
fix(datastore): cascade deleted for nested Has Many

### PR DESCRIPTION
**Pull request by iartemiev**
#### Description of changes

Fixes bug when deleting a record of a model with > 1 nested Has Many descendants
Ex:
```graphql
type Blog @model {
  id: String!
  posts: [Post] @hasMany
}

type Post @model {
  id: String!
  comments: [Comment] @hasMany
}

type Comment @model {
  id: String!
}
```
Behavior before the change in this PR:
1. `await DataStore.delete(Blog, 'some-id')`
2. `deleteTraverse` retrieves all `Post` records from the client database that are related to the Blog record and recursively calls `deleteTraverse`, passing the retrieved post records as POJOs ([link to code](https://github.com/aws-amplify/amplify-js/pull/10880/files#diff-75c76069840355a3f346f2f35c443018e4a5716ed128191602499c3b39935cfcR1076-R1080))
3. since `Post` has a Has Many relationship to `Comment`, `deleteTraverse` attempts to retrieve the related `Comment` records, but first has to extract PK metadata from the posts. Since the posts are POJOs, instead of model instances, they don't contain the necessary data and cause the error in the linked issue.

Note: If the child model does not have descendant relationships defined in the schema (e.g. if Post did not have a Has Many relationship to Comment) we wouldn't attempt to retrieve its descendants in step 3. so the bug would not show up with a relational graph of depth 2. Only 3+.

The change in this PR instantiates models from the records retrieved in step 2 and passes the models in the recursive call, which fixes the bug.

#### Issue #, if available

https://github.com/aws-amplify/amplify-js/issues/10866

#### Description of how you validated changes
* Red/green unit test
* Manual testing

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/commonAdapterTests.ts [hunk 1]
```diff
@@ -503,6 +506,35 @@ export function addCommonQueryTests({
...
+
+				const posts = await retrievedBlog!.posts.toArray();
+				expect(posts.length).toEqual(1);
+				expect(posts[0].id).toEqual(post.id);
+
+				await DataStore.delete(Blog, blog.id);
+
+				expect(await DataStore.query(Post, post.id)).toBeUndefined();
+			}
+		);
```
Comment (svidgen) [inline packages/datastore/__tests__/commonAdapterTests.ts:536]: What's the relevant difference between what this tests and [the test just above it](https://github.com/aws-amplify/amplify-js/pull/10880/files#diff-de9d5dae30ad1e8fd1b9539e2f4cfdd40c71894df2fbb38c779a409f3d8a4908R472-R507)?

Why was the existing test passing and this one red-green-able? Is the pre-existing test for this broken?

**Review by svidgen — COMMENTED** [review]
I think I see why this fixes the problem. But, I have one concern, noted in diff.

I'm also very curious whether this problem becomes simpler if handled outside the adapter. I'm thinking we should have an adapter-independent `DeleteTraversal` class or util function, and the adapters should just accept a list of items to delete (transactionally if supported, for the efficiency's sake if nothing else).

File: packages/datastore/__tests__/commonAdapterTests.ts (hunk 1 above)
Comment (iartemiev) [reply packages/datastore/__tests__/commonAdapterTests.ts:536]: The reason is the same as the Note in the PR description. You have to have at least 3 models in a relational graph for this bug to occur.

File: packages/datastore/__tests__/helpers.ts [hunk 2]
```diff
@@ -530,8 +536,7 @@ class FakeGraphQLService {
 			this.tables.set(model.name, new Map<string, any[]>());
 			let CPKFound = false;
 			for (const attribute of model.attributes || []) {
-				// Pretty sure the first key is the PK.
-				if (attribute.type === 'key') {
+				if (isModelAttributePrimaryKey(attribute)) {
```
Comment (svidgen) [inline packages/datastore/__tests__/helpers.ts:539]: Oh, thanks for fixing this!

**Comment by iartemiev** [comment]
> I'm also very curious whether this problem becomes simpler if handled outside the adapter. I'm thinking we should have an adapter-independent DeleteTraversal class or util function, and the adapters should just accept a list of items to delete (transactionally if supported, for the efficiency's sake if nothing else).

Abstracting out parts of the storage adapter implementations would avoid repetition, but the change in this PR would still apply.

File: packages/datastore/__tests__/helpers.ts (hunk 2 above)
Comment (iartemiev) [reply packages/datastore/__tests__/helpers.ts:539]: Np! The "key" difference with that util is that - it's not the first key attribute that's the PK. It's the key attribute without a `name` property

File: packages/datastore/__tests__/commonAdapterTests.ts (hunk 1 above)
Comment (svidgen) [reply packages/datastore/__tests__/commonAdapterTests.ts:536]: I'm probably being dense. But, how does the three-level-nesting matter if the test is only going two levels deep?

Comment (iartemiev) [reply packages/datastore/__tests__/commonAdapterTests.ts:536]: And with this unit test I added a `Blog` model that has a Has Many relationship to `Post`, which has a Has Many to `Comment`. Elsewhere in the unit tests we only test models with at most 1 descendant.

**Comment by svidgen** [comment]
> > I'm also very curious whether this problem becomes simpler if handled outside the adapter. I'm thinking we should have an adapter-independent DeleteTraversal class or util function, and the adapters should just accept a list of items to delete (transactionally if supported, for the efficiency's sake if nothing else).
> 
> Abstracting out parts of the storage adapter implementations would avoid repetition, but the change in this PR would still apply.

Woops. I posted a half-completed thought here! 🧠  Yes. I agree! And this is also outside of scope for the PR anyway. Seeing redundant changes and *yet another* "not relevant for SQLiteAdapter" test bumps the urgency level of some refactoring here. So, I wanted to plant the seed and looking for a +/- reaction to the idea. If it inspired a followup, I wouldn't be opposed. I also wouldn't oppose to adding an item to the backlog.

**Comment by iartemiev** [comment]
That's fair. I think we should merge this to unblock folks that are depending on it first (provided there aren't any issues with the PR). But I'll create a backlog task to refactor. It'll be a non-trivial effort because despite a lot of high level similarity, there are many adapter-specific code differences that we'll need to find a way to express elegantly.

File: packages/datastore/__tests__/commonAdapterTests.ts [hunk 3]
```diff
@@ -503,6 +506,35 @@ export function addCommonQueryTests({
 			}
 		);
 
+		(isSQLiteAdapter() ? test.skip : test)(
```
Comment (svidgen) [inline packages/datastore/__tests__/commonAdapterTests.ts:509]:
```suggestion
		/**
		 * Distinct from other hasMany delete cascade tests in that the model
		 * being deleted has *potential* grandchildren. Hence, delete traversals need
		 * to be capable of successfully querying the child for its children,
		 * which has seen a regression at least once!
		 * See: https://github.com/aws-amplify/amplify-js/issues/10866
		 */
		(isSQLiteAdapter() ? test.skip : test)(
```

**Review by svidgen — COMMENTED** [review]
Thanks for your patience explaining this to me. Does the suggested comment below on the new test make sense?

File: packages/datastore/__tests__/helpers.ts (hunk 2 above)
Comment (svidgen) [reply packages/datastore/__tests__/helpers.ts:539]: 😅

File: packages/datastore/__tests__/commonAdapterTests.ts (hunk 1 above)
Comment (svidgen) [reply packages/datastore/__tests__/commonAdapterTests.ts:536]: Thanks for explaining it and also making me feel less dumb about asking!

**Review by svidgen — APPROVED** [review]

**Review by david-mcafee — APPROVED** [review]
LGTM!

**Pull request merged**